File: Data Structures & Algorithms/subsets-ii/submission-0.cpp
class Solution {
private:
    void dfs(vector<int>& nums, int i, vector<vector<int>>& res, vector<int>& subset){
        if (i >= nums.size()){
            res.push_back(subset);
            return;
        }
        subset.push_back(nums[i]);
        dfs(nums, i + 1, res, subset);
        subset.pop_back();
        while (i + 1 < nums.size() && nums[i] == nums[i + 1]){
            i++;
        }
        dfs(nums, i + 1, res, subset);
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> subset;
        dfs(nums, 0, res, subset);
        return res;
    }
};

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        tracking = [False] * len(nums)
        res = []
        permutation = []
        def dfs(depth):
            if depth >= len(nums):
                res.append(permutation.copy())
            for i in range(len(nums)):
                if tracking[i] == False:
                    tracking[i] = True
                    permutation.append(nums[i])
                    dfs(depth + 1)
                    tracking[i] = False
                    permutation.pop()

        dfs(0)
        return res